### PR TITLE
ubuntu unattended setup and postinstall script

### LINKED ISSUE #193: [PostInstall] provide step to apply `/etc`-mods from image-setup
maybe we can add a step to the PostInstall script that parses the master/slave image setup config file for their `%post`-entries`and allow to execute those step-by-step (where confirmation is required before execution - and commands can be skipped if execution is not desired)

this way, we could update an already set up system with the updates we found in the meantime, e.g. `git-lfs`-setup or mimic-magic...

I'll try to provide a suggestion

### LINKED ISSUE #196: [image setup] create setup for 16.04. and Kinetic
323 asked whether `cob4-18` (base only) could be setup with 16.04. and Kinetic

in particular this requires a migration from `upstart` to `systemd`
https://wiki.ubuntuusers.de/systemd/

### PR DESCRIPTION
**Pull request by ipa-bnm**
fixes #196 
fixes #193

### CODE REVIEW

**Comment by ipa-fxm** [comment]
some observations when using the stick to set up `cob4-18` (323):
 - setting up udev rules via PostInstall failed - we had to remove `/dev/usb*` (@ipa-bnm, correct?) first, then re-run
 - `cob-start` still has a hardcoded `ping h1` which is not available on cob4-18 (base-only)
 - we still need to add a PostInstall step for running the `kickstart-robot.sh` - see #193

File: images_config/create_iso.sh [hunk 1]
```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/env bash
```
Comment (ipa-fxm) [inline images_config/create_iso.sh]: please add documentation for this in the README

File: images_config/kickstart/kickstart-robot.sh [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+}
+
+function UpgradeAptPackages {
+    printHeader "UpgradeAptPackages"
+    apt-get update
+    apt-get upgrade -y
+    apt-get dist-upgrade -y
+}
+
+# TODO: fxm
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: no longer a ToDo, right

File: images_config/kickstart/kickstart-robot.sh [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+function UpgradeKernel {
+    printHeader "UpgradeKernel"
+    if [ "$DISTRO" == "trusty" ]; then
+        apt-get install --install-recommends linux-generic-lts-xenial -y --allow
+    fi
+}
+
+function InstallHWEnableStacks {
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: if we use HW-enabled seed for 14.04 (like we do for 16.04), this step is not required here anymore

File: images_config/kickstart/kickstart-robot.sh [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+    printHeader "InstallROS"
+    if grep -q "deb http://packages.ros.org/ros/ubuntu $DISTRO main" /etc/apt/sources.list.d/ros-latest.list ; then
+        echo "Ros sources already setup. Skipping setup"
+    else
+        echo "deb http://packages.ros.org/ros/ubuntu $DISTRO main" > /etc/apt/sources.list.d/ros-latest.list
+        apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net:80 --recv-key 421C365BD9FF1F717815A3895523BAEEB01FA116
+    fi
+
+    apt-get update
+    apt-get install ros-kinetic-ros-base -y
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: make dual-distro compatible

File: images_config/kickstart/kickstart-robot.sh [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+            update-rc.d udev_cob.sh defaults
+    elif [ "$INSTALL_TYPE" == "slave" ]; then
+        wget -O /etc/udev/rules.d/99-gripper.rules https://raw.githubusercontent.com/ipa320/setup_cob4/master/udev_rules/99-gripper.rules
+    fi
+}
+
+function InstallGitLFS {
+    printHeader "InstallGitLFS"
+    curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash
+    apt-get install git-lfs
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: there is an additional step for git lfs to work (https://github.com/git-lfs/git-lfs/wiki/Installation)
`git lfs install`

....we need to make sure this is called before we start cloning lfs-enabled repos!

File: images_config/kickstart/kickstart-robot.sh [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+    apt-get purge modemmanager -y
+}
+
+function DisableUpdatePopup {
+    printHeader "DisableUpdatePopup"
+    sed -i 's/Prompt\=lts/Prompt\=never/g' /etc/update-manager/release-upgrades
+}
+
+# TODO: fxm ;)
+# Not functional under xenial
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: under trusty, this guaranteed that the slave pc's only booted with network set up...which was required to guarantee that nfs is running when logging into the mimic user....which was required for the GUI to be displayed properly via vlc

we need to find a solution for this under xenial, too!

File: images_config/kickstart/kickstart-robot.sh [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+        fi
+        if grep -q 'Acquire::http { Proxy "http://'$SERVERNAME':3142"; };' /etc/apt/apt.conf.d/01proxy ; then
+            echo "Proxy already in /etc/apt/apt.conf.d/01proxy, skipping InstallAptCacher"
+        else
+            echo 'Acquire::http { Proxy "http://'$SERVERNAME':3142"; };' >>  /etc/apt/apt.conf.d/01proxy
+        fi
+    fi
+}
+
+function InstallRealsense {
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: should we put this function together with https://github.com/ipa320/setup_cob4/pull/199/files#diff-49b47db197119e8cdf6e298fb8c6df5eR34?

File: images_config/kickstart/kickstart-robot.sh [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+        apt-get update
+        apt-get install realsense-uvcvideo -y
+        apt-get install ros-kinetic-librealsense -y
+        apt-get install ros-kinetic-realsense-camera -y
+    fi
+}
+
+function InstallCareOBot {
+    printHeader "InstallCareOBot"
+    apt-get install ros-kinetic-care-o-bot-robot -y
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: make dual-distro compatible

File: images_config/kickstart/kickstart-robot.sh [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+    printHeader "SetupGrubRecFail"
+    if grep -q GRUB_RECORDFAIL_TIMEOUT= /etc/default/grub ; then
+        echo "found GRUB_RECORD_FAIL flag already, skipping SetupGrubRecFail (update-grub call)"
+    else
+        echo "GRUB_RECORDFAIL_TIMEOUT=2" >> /etc/default/grub
+        update-grub
+    fi
+}
+
+#TODO: requires user input - No working solution found so far
```
Comment (ipa-fxm) [inline images_config/kickstart/kickstart-robot.sh]: isn't this fixed by https://github.com/ipa320/setup_cob4/pull/199/files#diff-448827c30f63f80de9bfc13be7a5bb3eR41?
if so, remove this function

File: images_config/kickstart/kickstart-robot.sh (hunk 6 above)
Comment (ipa-fxm) [reply images_config/kickstart/kickstart-robot.sh]: maybe `/etc/X11/Xsession.options` could be modified under `xenial` instead...

there is an option `failsafe-allowed` - that could be commented out...

File: images_config/kickstart/kickstart-robot.sh (hunk 2 above)
Comment (ipa-bnm) [reply images_config/kickstart/kickstart-robot.sh]: right

File: images_config/kickstart/kickstart-robot.sh (hunk 6 above)
Comment (ipa-bnm) [reply images_config/kickstart/kickstart-robot.sh]: As far as i understood, the failsafe boot has nothing to do with the graphical failsafe screen (e.g. "starting low graphics mode - graphics driver problem". Failsafe in this case has something to do with the runlevel. Runlevel 1 is recovery/single-mode and the script for 14.04 prevents recovery/single-mode.
We should observe if we still need it.

File: images_config/kickstart/kickstart-robot.sh (hunk 9 above)
Comment (ipa-bnm) [reply images_config/kickstart/kickstart-robot.sh]: if no user input is availabe, as it is on stick install, user input is not required for this to work.

**Pull request merged**